indexer/services/ender/src/lib/kafka-publisher.ts

```diff
@@ -43,34 +51,93 @@ export class KafkaPublisher {
-        this.tradeMessages.push(event.message);
-        break;
+        return this.tradeMessages;
       case KafkaTopics.TO_WEBSOCKETS_MARKETS:
-        this.marketMessages.push(event.message);
-        break;
+        return this.marketMessages;
       case KafkaTopics.TO_WEBSOCKETS_CANDLES:
-        this.candleMessages.push(event.message);
-        break;
+        return this.candleMessages;
       case KafkaTopics.TO_VULCAN:
-        this.vulcanMessages.push(event.message);
-        break;
+        return this.vulcanMessages;
       default:
         throw new Error('Invalid Topic');
     }
   }
 
   /**
-   * Sort trade events by block height, transaction index, and event index in ascending order,
-   * where the first trade event should be the earliest trade in the block.
+   * Sort subaccountMessages that represent fills by block height, transaction index,
+   * and event index in ascending order per order id. Only keep the subaccount message if
+   * it represents the last fill event per order id.
+   */
+  public retainLastFillEventsForSubaccountMessages() {
+    // Create a map to store the last fill event per order ID
+    const lastFillEvents: Record<string, AnnotatedSubaccountMessage> = {};
+    const nonFillEvents: AnnotatedSubaccountMessage[] = [];
+
+    this.subaccountMessages.forEach((message) => {
+      if (message.isFill && message.orderId) {
+        const orderId = message.orderId;
+        // If we haven't seen this order ID before or if the current message
+        // has a higher block height, update the lastFillEvents for this order ID
+        if (
+          !lastFillEvents[orderId] ||
+          this.compareMessages(message, lastFillEvents[orderId]) > 0
+        ) {
+          lastFillEvents[orderId] = message;
+        }
+      } else {
+        nonFillEvents.push(message);
+      }
+    });
+
+    this.subaccountMessages = Object.values(lastFillEvents)
+      .concat(nonFillEvents)
+      .map((annotatedMessage) => convertToSubaccountMessage(annotatedMessage));
+    this.sortEvents(KafkaTopics.TO_WEBSOCKETS_SUBACCOUNTS);
+  }
+
+  /** Helper function to compare two AnnotatedSubaccountMessages based on block height,
+   * transaction index, and event index.
+   */
+  private compareMessages(a: AnnotatedSubaccountMessage, b: AnnotatedSubaccountMessage) {
+    if (a.blockHeight === b.blockHeight) {
+      if (a.transactionIndex === b.transactionIndex) {
+        return a.eventIndex - b.eventIndex;
+      }
+      return a.transactionIndex - b.transactionIndex;
+    }
+    return Number(a.blockHeight) - Number(b.blockHeight);
+  }
+
+  /**
+   * Sort events by block height, transaction index, and event index in ascending order,
+   * where the first event should be the earliest event in the block.
    */
-  public sortTradeEvents() {
-    this.tradeMessages = this.tradeMessages.sort(
-      (a: SingleTradeMessage, b: SingleTradeMessage) => {
+  public sortEvents(kafkaTopic: KafkaTopics) {
```

Comment:
nit: can this be made a generic function that takes in `OrderedMessage` and sorts them and returns a sorted list rather than checking if the topic is sortable?

Reply:
done